collection-scripts/gather_priority_and_fairness

```diff
@@ -0,0 +1,60 @@
+#!/bin/bash
+BASE_COLLECTION_PATH="/must-gather"
+NAMESPACE_PATH="namespaces/openshift-kube-apiserver"
+POD_PATH="kube-apiserver/kube-apiserver/api_priority_and_fairness"
+APF_PATH="${BASE_COLLECTION_PATH}/${NAMESPACE_PATH}"
+
+# If the cluster version is < 4.6 skip this test
+
+# If cluster version is >=4.6 then the COMPARE will always == 4.6
+#  For example cluster version 4.7:
+#  VERSION=4.7
+#  COMPARE=$(echo -e "4.6\n${VERSION}" | sort -V | head -n1)
+#  echo $COMPARE
+#  4.6
+
+# If cluster version is <4.6 then the COMPARE will be the always the cluster version.
+#  VERSION=4.4
+#  COMPARE=$(echo -e "4.6\n${VERSION}" | sort -V | head -n1)
+#  echo $COMPARE
+#  4.4
+
+VERSION=$(oc get clusterversion version -o jsonpath='{.status.desired.version}')
+COMPARE=$(echo -e "4.6\n${VERSION}" | sort -V | head -n1)
+if [[ ${COMPARE} != "4.6" ]]; then
+  exit 0
+fi
+
+# Fetches the POD_NAME where app label is apiserver.
+# Example:
+#  oc get po -n openshift-kube-apiserver -l apiserver -o jsonpath='{range .items[*]}{.metadata.name}{"\n"}'
+#  kube-apiserver-master-0.example.com 192.168.9.43 6443
+#  kube-apiserver-master-1.example.com 192.168.13.112 6443
+#  kube-apiserver-master-2.example.com 192.168.10.112 6443
+KUBE_API_SERVERS=$(oc get po -n openshift-kube-apiserver -l apiserver -o jsonpath='{range .items[*]}{.metadata.name}{"\n"}')
+
+for API_SERVER in ${KUBE_API_SERVERS}
+do
+  OUTPUT_DIR="${APF_PATH}/pods/${API_SERVER}/${POD_PATH}"
+  mkdir -p "${OUTPUT_DIR}"
+
+  # Fetches the POD_IP from the apiserver pod.
+  # Example:
+  #  oc get po -n openshift-kube-apiserver kube-apiserver-master-0.example.com -o jsonpath='{.status.podIP}'
+  #  192.168.9.43
+  POD_IP=$(oc get po -n openshift-kube-apiserver "${API_SERVER}" -o jsonpath='{.status.podIP}')
+
+  # Fetches the POD_PORT where container name is kube-apiserver.
+  # Example:
+  #  oc get po -n openshift-kube-apiserver kube-apiserver-master-0.example.com -o jsonpath='{range .spec.containers[?(.name=="kube-apiserver")]}{.ports[0].hostPort}'
+  #  6443
+  POD_PORT=$(oc get po -n openshift-kube-apiserver "${API_SERVER}" -o jsonpath='{range .spec.containers[?(.name=="kube-apiserver")]}{.ports[0].hostPort}')
+
+  oc get --raw https://"${POD_IP}":"${POD_PORT}"/debug/api_priority_and_fairness/dump_priority_levels \
```

Comment:
what happens if the request fails, does it get retried by must-gather?

in my opinion, `must-gather` is usually run when the cluster is under some sort of duress. We have three requests here, if one fails will the entire must-gather process abort?

Reply:
No retries have been implemented. I will take a look at it today. Thanks

Reply:
if we want to do retry, it can be done in a follow up PR.

Reply:
Is this really where we want to do this? 
Should this go in: https://github.com/openshift/oc/blob/master/pkg/cli/admin/inspect/pod.go#L161-L180 which already gets called when we collect the clusteroperators (related resources) with https://github.com/openshift/must-gather/blob/master/collection-scripts/gather#L19 and https://github.com/openshift/must-gather/blob/master/collection-scripts/gather#L43

Reply:
I don't think we should retry. In fact, a request will be already retried by `oc`.
We should make sure that the `must-gather` won't fail if we can not get some data. AFAIK it already works like this.

@tkashem thoughts?

Reply:
> Should this go in: https://github.com/openshift/oc/blob/master/pkg/cli/admin/inspect/pod.go#L161-L180 which

oh, I wasn't aware of it. It looks really good. Though I have never seen debug data attached to a must-gather.

Reply:
ah, so it gets collected on `oc adm inspect` not `oc adm must-gather`. It seems that `inspect` is for collecting debugging data

Reply:
@sferich888 

I could use that debugEndpoints instead and it's pretty much better place to put it into. But how does the `oc adm inspect` work without token?

All my endpoint collections using `oc adm inspect` were skipped because Token is not empty.
https://github.com/openshift/oc/blob/master/pkg/cli/admin/inspect/pod.go#L58-L62

So it never got to:
https://github.com/openshift/oc/blob/master/pkg/cli/admin/inspect/pod.go#L72-L74

Which calls debug Endpoints collection:
https://github.com/openshift/oc/blob/f749845d71de73620f5e66553e0dd5abeb7efede/pkg/cli/admin/inspect/pod.go#L112-L113

That calls: https://github.com/openshift/oc/blob/f749845d71de73620f5e66553e0dd5abeb7efede/pkg/cli/admin/inspect/pod.go#L168-L177

Reply:
So to me it seems that first we exit out of `gatherContainerInfo` before collecting debug endpoints because we have a token.

```
	if len(o.RESTConfig.BearerToken) > 0 {
		// token authentication is vulnerable to replays if the token is sent to a potentially untrustworthy source.
		klog.V(1).Infof("        Skipping container endpoint collection for pod %q container %q: Using token authentication\n", pod.Name, container.Name)
		return nil
	}
```

And then in `gatherContainerDebug` which we never get into, actually expect the token?

```
	for _, endpoint := range endpoints {
		// we need a token in order to access the /debug endpoint
		result, err := o.podUrlGetter.Get("/debug/pprof/"+endpoint, pod, o.restConfig, debugPort)
		if err != nil {
			return err
		}
		if err := o.fileWriter.WriteFromSource(path.Join(destDir, endpoint), &TextWriterSource{Text: result}); err != nil {
			return err
		}
	}
```

Reply:
I have just run `oc adm inspect clusteroperators` on my local cluster and here is what I have found for `kas`

```
ls openshift-kube-apiserver/pods/kube-apiserver-ip-10-0-149-59.ec2.internal/kube-apiserver/kube-apiserver
healthz      heap         logs         metrics.json profile      trace        version.json
```